src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,548 @@
+        """
+        Initialises work spaces.
+        For every dimension in the feature space, creates a minimum and a maximum work range.
+        """
+        for i in range(self.dimensions):
+            sq = random.uniform(0, 1)
+            min_element = sq - 2 * max(sq, 1 - sq)
+            max_element = sq + 2 * max(sq, 1 - sq)
+            try:
+                self.min[i] = min_element
+            except IndexError:
+                self.min.append(min_element)
+            try:
+                self.max[i] = max_element
+            except IndexError:
+                self.max.append(max_element)
+
+    def build_trees(self):
+        """
+        Initialises ensemble.
+
+        """
+        for i in range(self.n_estimators):
+            self.initialise_work_space()
+            tree = HalfSpaceTree(self.depth, self.dimensions, self.size_limit)
+            tree.root = tree.build_tree(self.min, self.max)
+            self.ensemble.append(tree)
+
+    def update_mass(self, X, boolean):
+        """
+        Populates mass profiles for every tree defined in the ensemble.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+
+        boolean: boolean, True or False
+            True to update reference mass, False to update latest mass
+        """
+        for i in range(self.n_estimators):
+            self.ensemble[i].update_mass(X, boolean)
+
+    def update_models(self):
+        """
+        Updates the mass profile of every tree in the ensemble.
+
+        """
+        for i in range(self.n_estimators):
+            tree = self.ensemble[i]
+            tree.update_model(tree.root)
+
+    def set_is_learning_phase_on(self, boolean):
+        """
+        Sets learning phase in each tree defined in the ensemble.
+
+        Parameters
+        ----------
+        boolean: Boolean
+
+        """
+        self.is_learning_phase_on = boolean
+        for i in range(self.n_estimators):
+            self.ensemble[i].is_learning_phase_on = boolean
+
+
+class HalfSpaceTree:
+
+    def __init__(self, max_depth, dimensions, size_limit):
+        """
+        Half Space Tree
+
+        Parameters
+            ----------
+        dimensions: int, required
+            The dimensionality of the stream.
+
+        max_depth: int, optional (default=15)
+            The maximum depth of the trees in the ensemble.
+            'maxDepth' in the original paper.
+
+        size_limit: int, optional (default=50)
+            The minimum mass required in a node (as a fraction of the window size) to calculate the anomaly score.
+            'sizeLimit' in the original paper.
+        """
+        super().__init__()
+        self.max_depth = max_depth
+        self.dimensions = dimensions
+        self.root = HalfSpaceTreeNode()
+        self.size_limit = size_limit
+        self.is_learning_phase_on = True
+
+    def predict_proba(self, X, max_score):
```

Comment:
I am not sure about why this functionality is now between `HalfSpaceTrees` and `HalfSpaceTree`. I will review the code once we focus on ``HalfSpaceTrees`, just leaving a comment for the moment.